File: app/services/alert_engine.py
"""Motor de alertas. Corre después de cada transacción nueva."""
from datetime import datetime, date
from sqlalchemy.orm import Session
from ..models import Transaction, Alert, User

def _capitalize(s: str) -> str:
    return s.capitalize()

async def run_alert_engine(user_id: int, db: Session):
    now = date.today()
    month = now.strftime("%Y-%m")
    days_in_month = (date(now.year, now.month % 12 + 1, 1) - date(now.year, now.month, 1)).days if now.month < 12 \
        else (date(now.year + 1, 1, 1) - date(now.year, now.month, 1)).days
    days_passed = now.day
    days_remaining = days_in_month - days_passed

    user = db.query(User).filter_by(id=user_id).first()
    if not user or not user.monthly_income:
        return

    income = user.monthly_income
    limits = {
        "necesidades": income * user.necesidades_pct / 100,
        "gustos": income * user.gustos_pct / 100,
        "ahorro": income * user.ahorro_pct / 100,
    }

    txs = db.query(Transaction).filter(
        Transaction.user_id == user_id,
        Transaction.month == month
    ).all()

    spent = {}
    for cat in limits:
        spent[cat] = sum(t.amount for t in txs if t.category == cat)

    one_day_ago = datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)

    for cat, limit in limits.items():
        s = spent.get(cat, 0)
        pct = s / limit if limit > 0 else 0

        # Threshold 90%
        if pct >= 0.9:
            if not _alert_exists(db, user_id, "threshold", cat, one_day_ago):
                remaining = limit - s
                msg = f"{_capitalize(cat)} casi en el límite. Te quedan ${remaining:,.0f} para {days_remaining} días."
                ai_msg = await _get_ai_advice(cat, s, limit, txs, db) if _claude_available() else None
                db.add(Alert(user_id=user_id, type="threshold", category=cat,
                             message=msg, ai_advice=ai_msg, severity="critical"))

        elif pct >= 0.75:
            if not _alert_exists(db, user_id, "threshold", cat, one_day_ago):
                msg = f"Ojo, venís rápido con {cat}. Estás al {int(pct*100)}% y quedan {days_remaining} días."
                ai_msg = await _get_ai_advice(cat, s, limit, txs, db) if _claude_available() else None
                db.add(Alert(user_id=user_id, type="threshold", category=cat,
                             message=msg, ai_advice=ai_msg, severity="warning"))

        # Proyección
        elif pct < 0.75 and days_passed > 0:
            daily_rate = s / days_passed
            projection = s + daily_rate * days_remaining
            if projection > limit and not _alert_exists(db, user_id, "projection", cat, one_day_ago):
                msg = f"Si seguís así, {cat} va a pasar el límite antes de fin de mes."
                db.add(Alert(user_id=user_id, type="projection", category=cat,
                             message=msg, severity="warning"))

    db.commit()

def _alert_exists(db: Session, user_id: int, type_: str, category: str, since) -> bool:
    return db.query(Alert).filter(
        Alert.user_id == user_id,
        Alert.type == type_,
        Alert.category == category,
        Alert.is_read == False,
        Alert.created_at >= since
    ).first() is not None

def _claude_available() -> bool:
    from ..config import get_settings
    return get_settings().claude_enabled

async def _get_ai_advice(category: str, spent: float, limit: float, txs: list, db: Session) -> str:
    from ..config import get_settings
    settings = get_settings()
    if not settings.claude_enabled:
        return ""

    import anthropic
    cat_txs = [(t.merchant, t.amount) for t in txs if t.category == category]
    top = sorted(cat_txs, key=lambda x: x[1], reverse=True)[:5]
    top_str = "\n".join(f"- {m}: ${a:,.0f}" for m, a in top)

    prompt = f"""Sos un coach financiero argentino, cercano y directo.

El usuario lleva ${spent:,.0f} gastados en {category} de un límite de ${limit:,.0f}.
Sus últimos gastos en esta franja:
{top_str}

Escribí UN consejo concreto y específico (máximo 2 orsentences, tono informal rioplatense).
No des consejos genéricos. Hacé referencia a los gastos reales.
No uses emojis. No uses "hola". Arrancá directo al punto."""

    try:
        import anthropic as ant
        client = ant.Anthropic(api_key=settings.claude_api_key)
        msg = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=150,
            messages=[{"role": "user", "content": prompt}]
        )
        return msg.content[0].text.strip()
    except Exception:
        return ""

File: services/transactions-service/src/services/alertEngine.js
/** Motor de alertas. Corre después de cada transacción nueva. */
import { prisma } from "@guita-coach/db";
import { daysInMonth, settings } from "@guita-coach/shared";
function capitalize(s) {
  return s.charAt(0).toUpperCase() + s.slice(1);
}
async function alertExists(userId, type, category, since) {
  const found = await prisma.alert.findFirst({
    where: {
      userId,
      type,
      category,
      isRead: false,
      createdAt: {
        gte: since
      }
    }
  });
  return Boolean(found);
}
async function getAiAdvice(category, spent, limit, txs) {
  if (!settings.claudeEnabled) return "";
  try {
    const catTxs = txs.filter(t => t.category === category).map(t => [t.merchant, t.amount]);
    const top = [...catTxs].sort((a, b) => (b[1] ?? 0) - (a[1] ?? 0)).slice(0, 5);
    const topStr = top.map(([m, a]) => `- ${m}: $${(a ?? 0).toLocaleString("es-AR")}`).join("\n");
    const prompt = `Sos un coach financiero argentino, cercano y directo.

El usuario lleva $${spent.toLocaleString("es-AR")} gastados en ${category} de un límite de $${limit.toLocaleString("es-AR")}.
Sus últimos gastos en esta franja:
${topStr}

Escribí UN consejo concreto y específico (máximo 2 oraciones, tono informal rioplatense).
No des consejos genéricos. Hacé referencia a los gastos reales.
No uses emojis. No uses "hola". Arrancá directo al punto.`;
    const r = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "x-api-key": settings.claudeApiKey,
        "anthropic-version": "2023-06-01"
      },
      body: JSON.stringify({
        model: "claude-sonnet-4-6",
        max_tokens: 150,
        messages: [{
          role: "user",
          content: prompt
        }]
      })
    });
    if (!r.ok) return "";
    const data = await r.json();
    return (data?.content?.[0]?.text ?? "").trim();
  } catch {
    return "";
  }
}
export async function runAlertEngine(userId) {
  const now = new Date();
  const month = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
  const days = daysInMonth(now.getFullYear(), now.getMonth() + 1);
  const daysPassed = now.getDate();
  const daysRemaining = days - daysPassed;
  const user = await prisma.user.findUnique({
    where: {
      id: userId
    }
  });
  if (!user || !user.monthlyIncome) return;
  const income = user.monthlyIncome;
  const limits = {
    necesidades: income * user.necesidadesPct / 100,
    gustos: income * user.gustosPct / 100,
    ahorro: income * user.ahorroPct / 100
  };
  const txs = await prisma.transaction.findMany({
    where: {
      userId,
      month
    }
  });
  const spent = {};
  for (const cat of Object.keys(limits)) {
    spent[cat] = txs.filter(t => t.category === cat).reduce((s, t) => s + (t.amount ?? 0), 0);
  }
  const oneDayAgo = new Date(now);
  oneDayAgo.setHours(0, 0, 0, 0);
  for (const [cat, limit] of Object.entries(limits)) {
    const s = spent[cat] || 0;
    const pct = limit > 0 ? s / limit : 0;
    if (pct >= 0.9) {
      if (!(await alertExists(userId, "threshold", cat, oneDayAgo))) {
        const remaining = limit - s;
        const msg = `${capitalize(cat)} casi en el límite. Te quedan $${remaining.toLocaleString("es-AR")} para ${daysRemaining} días.`;
        const aiMsg = await getAiAdvice(cat, s, limit, txs);
        await prisma.alert.create({
          data: {
            userId,
            type: "threshold",
            category: cat,
            message: msg,
            aiAdvice: aiMsg,
            severity: "critical"
          }
        });
      }
    } else if (pct >= 0.75) {
      if (!(await alertExists(userId, "threshold", cat, oneDayAgo))) {
        const msg = `Ojo, venís rápido con ${cat}. Estás al ${Math.round(pct * 100)}% y quedan ${daysRemaining} días.`;
        const aiMsg = await getAiAdvice(cat, s, limit, txs);
        await prisma.alert.create({
          data: {
            userId,
            type: "threshold",
            category: cat,
            message: msg,
            aiAdvice: aiMsg,
            severity: "warning"
          }
        });
      }
    } else if (daysPassed > 0) {
      const dailyRate = s / daysPassed;
      const projection = s + dailyRate * daysRemaining;
      if (projection > limit && !(await alertExists(userId, "projection", cat, oneDayAgo))) {
        const msg = `Si seguís así, ${cat} va a pasar el límite antes de fin de mes.`;
        await prisma.alert.create({
          data: {
            userId,
            type: "projection",
            category: cat,
            message: msg,
            severity: "warning"
          }
        });
      }
    }
  }
}
